src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexerSender.java

```diff
@@ -0,0 +1,224 @@
+// Copyright 2015 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.shell.Subprocess;
+import com.google.devtools.build.lib.shell.SubprocessBuilder;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.Semaphore;
+
+/**
+ * An intermediate worker that send and receive request from the actual worker
+ */
+public class WorkerMultiplexerSender extends Thread {
```

Comment:
I know we talked briefly about combining the sender and receiver, and just having the `WorkerProxy`s write directly to the process's `stdin` stream through a synchronized method on the unified multiplexer. Do you think that would make things less complex overall?

Reply:
I believe it is doable. I can make the change and see if we get any issue.